maxima/g0/neutrals/sigma-cx-calc.mac

```diff
@@ -0,0 +1,141 @@
+  for i : 1 thru length(expr) do (	
+    if expr[i] # 0.0 then (
+      ep : string(expr[i]),
+      ep : ssubst("log(","ln(",ep),	
+      printf(fh, "  ~a = ~a; ~%", lhs[i-1], ep)	
+    )
+  )
+)$
+
+writeCExprsC(lhs, rhs) := block([expr],
+  expr : float(expand(rhs)),
+  if expr # 0.0 then (
+     ep : string(expr),
+     printf(fh, "  ~a = ~a; ~%", lhs, ep)	
+  )
+)$
+
+/* vSigmaCX for charge exchange cross section */
+calcSigmaCXcellAv(fh, funcNm, cdim, vdim, basisFun, pMax) := block([],
+  kill(varsC, varsP, basisC, basisP),
+  modNm : sconcat("basis-precalc/basis", basisFun, cdim, "x", vdim, "v"),
+  load(modNm),
+  for p : 1 thru pMax do (
+    bP : basisP[p],
+    print(basisFun, cdim, vdim, p, length(basisP)),
+    bC : basisC[p],
+    NC : length(bC),
+
+    printf(fh, " ~%"),
+    printf(fh, "GKYL_CU_DH static inline ~%"),
+    printf(fh, "double ~a_p~a(const double a, const double b, double vt_sq_ion_min, double vt_sq_neut_min, const double *m0, const double *prim_vars_ion, const double *prim_vars_neut, double* GKYL_RESTRICT v_sigma_cx) ~%{ ~%", funcNm, p),
+    printf(fh, "  // a               constant in fitting function. ~%"),
+    printf(fh, "  // b               constant in fitting function. ~%"),
+    printf(fh, "  // m0[~a]:         neutral particle density. ~%", NC),
+    printf(fh, "  // prim_vars_ion[~a]:   ion prim vars. ~%", NC*(vdim+1)),
+    printf(fh, "  // prim_vars_neut[~a]:  neut prim vars. ~%", NC*(vdim+1)),
+    printf(fh, "  // v_sigma_cx:          cell ave cross section fitting eqn. ~%"),
+    printf(fh, " ~%"),
+
+    m0_e : doExpand1(m0,bC),
+    m0_av :  innerProd(varsC,1.0,1.0,m0_e)/innerProd(varsC,1.0,1.0,1.0),
+    declareWriteCExprsC(m0_neut_av, m0_av),
+
+    printf(fh, "  const double *u_ion = &prim_vars_ion[~a]; ~%", 0),
+    printf(fh, "  const double *vt_sq_ion = &prim_vars_ion[~a]; ~%", NC*vdim),
+    printf(fh, "  const double *u_neut = &prim_vars_neut[~a]; ~%", 0),	
+    printf(fh, "  const double *vt_sq_neut = &prim_vars_neut[~a]; ~%", NC*vdim),
+    printf(fh, " ~%"),
+
+    u_ion_av_c : subst(makelist(u_ion[i]=u_ion[i+a0],i,0,NC-1),innerProd(varsC,1.0,1.0,doExpand1(u_ion,bC))/innerProd(varsC,1.0,1.0,1.0)),
+    u_neut_av_c : subst(makelist(u_neut[i]=u_neut[i+a0],i,0,NC-1),innerProd(varsC,1.0,1.0,doExpand1(u_neut,bC))/innerProd(varsC,1.0,1.0,1.0)),
+
+    u_ion_av : makelist(0,i,1,vdim),
+    u_neut_av : makelist(0,i,1,vdim),
+    for vi : 1 thru vdim do (
+      u_ion_av[vi] : fullratsimp(subst(a0=(vi-1)*NC,u_ion_av_c)),
+      u_neut_av[vi] : fullratsimp(subst(a0=(vi-1)*NC,u_neut_av_c))	
+    ),
+
+    vt_sq_ion_av_c : innerProd(varsC,1.0,1.0, doExpand1(vt_sq_ion,bC))/innerProd(varsC,1.0,1.0,1.0),
+    vt_sq_neut_av_c : innerProd(varsC,1.0,1.0, doExpand1(vt_sq_neut,bC))/innerProd(varsC,1.0,1.0,1.0),
+
+    declareWriteCExprsC(vt_sq_ion_av, vt_sq_ion_av_c),
+    declareWriteCExprsC(vt_sq_neut_av, vt_sq_neut_av_c),
+
+    /* Limit how low the temperature can get (set by the velocity grid). */
```

Comment:
In this code you turn off cx if vtsq < 0. I wonder if we should instead cap cx, by setting vtsq_av = vtsq_min even if vtsq<0. Not sure which is best. I guess if the ion temperature is too low (vtsq<0), you wouldn't want to cool the ions further by letting them CX, but it'd be fine for the neutrals because it'd heat them up, right?